### PR TITLE
#2292 parent path differences

### LINKED ISSUE #2292: "Parent path" differences are not shown in "Comparison"

### PR DESCRIPTION
**Pull request by benjaperez1983**
Fixes #2292 

# Description

Parent path differences now showing on comparison

## Type of change

Please mark relevant options with an `x` in the brackets.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires documentation changes (link at least one [user](https://github.com/Open-Systems-Pharmacology/docs) or [developer](https://github.com/Open-Systems-Pharmacology/developer-docs) documentation issue):
- [ ] Algorithm update - updates algorithm documentation/questions/answers etc.
- [ ] Other (please describe):

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Integration tests
- [ ] Unit tests
- [x] Manual tests 
- [ ] No tests required

# Reviewer checklist

Mark everything that needs to be checked before merging the PR.

- [ ] Check if the code is well documented
- [ ] Check if the behavior is what is expected
- [ ] Check if the code is well tested
- [ ] Check if the code is readable and well formatted
- [ ] Additional checks (document below if any)
- [ ] Check if documentation update issue(s) are created if the option `This change requires a documentation update` above is selected

# Screenshots (if appropriate):

# Questions (if appropriate):

### CODE REVIEW

**Review by msevestre — CHANGES_REQUESTED** [review]
Unit test missing
Bug fix=> unit test. No exception to the rule

**Comment by benjaperez1983** [comment]
> Unit test missing Bug fix=> unit test. No exception to the rule

Well, it was not a bug in fact, I`ve marked it as that but it is a new feature.
Having said that, challenge accepted, it should have it`s tests.

IMPORTANT: 
I added one file and renamed an existing, since the naming of the spec class was not accurate at all.

File: tests/OSPSuite.Core.IntegrationTests/DiffBuilders/ContainerDiffBuilderSpecs.cs [hunk 1]
```diff
@@ -1,99 +1,46 @@
...
-         var container1 = new Container {Name = "O1"};
-         container1.Add(new Parameter {BuildMode = ParameterBuildMode.Global, Name = "P1"});
-
-         var container2 = new Container {Name = "O2"};
-         container2.Add(new Parameter {BuildMode = ParameterBuildMode.Local, Name = "P1"});
-
-         _object1 = container1;
-         _object2 = container2;
-      }
+         var entityDiffBuilder = A.Fake<EntityDiffBuilder>();
```
Comment (msevestre) [inline tests/OSPSuite.Core.IntegrationTests/DiffBuilders/ContainerDiffBuilderSpecs.cs:18]: You are moving from Integration tests actually testing the real behavior to using fake. I don't understand the reasoning. In fact, why are you not using concern_for_ObjectComparer ? This does everything for you. You just have to set _object1 and object2 in all your subtests and everything is done for you. RIght now, you are reinvitening the wheel

File: tests/OSPSuite.Core.IntegrationTests/DiffBuilders/ContainerObjectComparerSpecs.cs [hunk 2]
```diff
@@ -0,0 +1,99 @@
+﻿using System.Linq;
+using OSPSuite.BDDHelper;
+using OSPSuite.BDDHelper.Extensions;
+using OSPSuite.Core.Domain;
+
+namespace OSPSuite.Core.DiffBuilders
+{
+   public class When_comparing_two_container_having_different_base_properties_and_container_properties : concern_for_ObjectComparer
```
Comment (msevestre) [inline tests/OSPSuite.Core.IntegrationTests/DiffBuilders/ContainerObjectComparerSpecs.cs:8]: Why did you rename the file. We are testing the behavior of ContainerDiff. We do not have an object called ContainerObjectComparer

**Review by msevestre — CHANGES_REQUESTED** [review]
I do not udnerstand the proposed changes
We just need to add one test to what we had before making sure the diff for ParentPath is reported. Do not use Fake unless absolutely required.

File: tests/OSPSuite.Core.IntegrationTests/DiffBuilders/ContainerObjectComparerSpecs.cs (hunk 2 above)
Comment (benjaperez1983) [reply tests/OSPSuite.Core.IntegrationTests/DiffBuilders/ContainerObjectComparerSpecs.cs:8]: you are right about that, but this was called "ContainterDiffBuilderSpecs", we do have a ContainterDiffBuilder class abd this tests are testing the "ContainerDiffBuilder" at all.

Since the concern class is concern_for_ObjectComparer : ContextForIntegration<IObjectComparer> , I thought it should be named after that.
Let me know what do you think

File: tests/OSPSuite.Core.IntegrationTests/DiffBuilders/ContainerDiffBuilderSpecs.cs (hunk 1 above)
Comment (benjaperez1983) [reply tests/OSPSuite.Core.IntegrationTests/DiffBuilders/ContainerDiffBuilderSpecs.cs:18]: it does not in fact, since the change I did is on the ContainerDiffBuilder Class, which was not being tested at all.
The tests (even when the class was called "ContainerDiffBuilderSpecs" is testing the "ObjectComparer" class, not the ContainerDiffBuilder.
And since my changes are on the ContainerDiffBuilder class, I created this tests and changed the previous class name as well.

File: tests/OSPSuite.Core.IntegrationTests/DiffBuilders/ContainerObjectComparerSpecs.cs (hunk 2 above)
Comment (msevestre) [reply tests/OSPSuite.Core.IntegrationTests/DiffBuilders/ContainerObjectComparerSpecs.cs:8]: this is not correct. We are testing the behavior of the ContainerDiff Builder implictely by making the diff of two contaienrs. This is how it goes here. Instead of always duplicating the logic, we created one base class, and we always create in the context the two objects to compare. I want to test the ContainerDiffBuilder, I create two containers  etc..

Because it's integration tests, we are testing the whole system as a whole but we are focusing here on Container. So in our case, we ARE Testing container diff builder by creating TWO containers. You just need to update the context to have two different ParentPath and that's it

File: tests/OSPSuite.Core.IntegrationTests/DiffBuilders/ContainerDiffBuilderSpecs.cs [hunk 3]
```diff
@@ -96,4 +98,23 @@ public void should_have_set_the_parent_to_the_common_3rd_level()
...
+
+         _object1 = new Container { Name = "O1", ParentPath = new ObjectPath("A", "B") };
+         _object2 = new Container { Name = "O1", ParentPath = new ObjectPath("A", "C") };
+      }
+
+      [Observation]
+      public void should_return_a_report_containing_the_container_path_difference()
+      {
+         _report.Count().ShouldBeEqualTo(1);
+         Assert.IsTrue(_report.First().Description.StartsWith("Parent Paths are not equal"));
```
Comment (msevestre) [inline tests/OSPSuite.Core.IntegrationTests/DiffBuilders/ContainerDiffBuilderSpecs.cs:117]: ok this is it. One last change. WE never use the ASsert but use our BDD stuff
can you do something along the line of _report.First().....ShouldBeEqualTo 
and remove the ref to NuNit at the top. 
also is there a constant that you can use instead of the magic string Parent Paths are not equal? Are we typically comparing to strings in the tests ?

Comment (benjaperez1983) [reply tests/OSPSuite.Core.IntegrationTests/DiffBuilders/ContainerDiffBuilderSpecs.cs]: Sure thing, just did          _report.First().Description.StartsWith("Parent Paths are not equal").ShouldBeTrue();
which is in line with our asserts.

Great catch mate!
